File: src/cpp/Source/Chunks/Shex/Tokens/TypedUnorderedAccessViewDeclarationToken.cpp
#include "PCH.h"
#include "TypedUnorderedAccessViewDeclarationToken.h"

#include "Decoder.h"

using namespace std;
using namespace SlimShader;

shared_ptr<TypedUnorderedAccessViewDeclarationToken> TypedUnorderedAccessViewDeclarationToken::Parse(BytecodeReader& reader)
{
	auto token0 = reader.ReadUInt32();
	auto coherency = DecodeValue<UnorderedAccessViewCoherency>(token0, 16, 16);
	auto operand = Operand::Parse(reader, DecodeValue<OpcodeType>(token0, 0, 10));
	auto result = shared_ptr<TypedUnorderedAccessViewDeclarationToken>(new TypedUnorderedAccessViewDeclarationToken(coherency, operand));
	result->_resourceDimension = DecodeValue<ResourceDimension>(token0, 11, 15);
	result->_returnType = ResourceReturnTypeToken::Parse(reader);
	return result;
}

ResourceDimension TypedUnorderedAccessViewDeclarationToken::GetResourceDimension() const { return _resourceDimension; }
const ResourceReturnTypeToken& TypedUnorderedAccessViewDeclarationToken::GetReturnType() const { return _returnType; }

void TypedUnorderedAccessViewDeclarationToken::Print(std::ostream& out) const
{
	out << GetTypeDescription() << " " << GetOperand();
}

TypedUnorderedAccessViewDeclarationToken::TypedUnorderedAccessViewDeclarationToken(UnorderedAccessViewCoherency coherency, Operand operand)
	: UnorderedAccessViewDeclarationTokenBase(coherency, operand)
{

}

File: src/SlimShader/Chunks/Shex/Tokens/TypedUnorderedAccessViewDeclarationToken.cs
using SlimShader.Util;

namespace SlimShader.Chunks.Shex.Tokens
{
	/// <summary>
	/// Typed Unordered Access View Declaration
	///
	/// OpcodeToken0:
	///
	/// [10:00] D3D11_SB_OPCODE_DCL_UNORDERED_ACCESS_VIEW_TYPED
	/// [15:11] D3D10_SB_RESOURCE_DIMENSION
	/// [16:16] D3D11_SB_GLOBALLY_COHERENT_ACCESS or 0 (LOCALLY_COHERENT)
	/// [23:17] Ignored, 0
	/// [30:24] Instruction length in DWORDs including the opcode token.
	/// [31]    0 normally. 1 if extended operand definition, meaning next DWORD
	///         contains extended operand description.  This dcl is currently not
	///         extended.
	///
	/// OpcodeToken0 is followed by 2 operands:
	/// (1) an operand, starting with OperandToken0, defining which
	///     u# register (D3D11_SB_OPERAND_TYPE_UNORDERED_ACCESS_VIEW) is being declared.
	/// (2) a Resource Return Type token (ResourceReturnTypeToken)
	/// </summary>
	public class TypedUnorderedAccessViewDeclarationToken : UnorderedAccessViewDeclarationTokenBase
	{
		public ResourceDimension ResourceDimension { get; private set; }
		public ResourceReturnTypeToken ReturnType { get; private set; }

		public static TypedUnorderedAccessViewDeclarationToken Parse(BytecodeReader reader)
		{
			var token0 = reader.ReadUInt32();

			return new TypedUnorderedAccessViewDeclarationToken
			{
				ResourceDimension = token0.DecodeValue<ResourceDimension>(11, 15),
				Coherency = token0.DecodeValue<UnorderedAccessViewCoherency>(16, 16),
				Operand = Operand.Parse(reader, token0.DecodeValue<OpcodeType>(0, 10)),
				ReturnType = ResourceReturnTypeToken.Parse(reader)
			};
		}
	}
}